
library(dada2)
library(ShortRead)
library(Biostrings)
library(tidyverse)


FWD <- "TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGTGYCAGCMGCCGCGGTAA"
REV <- "GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGGACTACNVGGGTWTCTAAT"

allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna),
        RevComp = reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}

FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)


fn <- list.files("./", pattern="fastq", full.names=TRUE)

primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}

#for (i in 1:10) { checkPrimers <- rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = file.path(fn[[i]])),
    FWD.ReverseReads = sapply(REV.orients, primerHits, fn = file.path(fn[[i]]))); print (checkPrimers)}

for (i in 1:10) {
  checkPrimers <- rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = file.path(fn[[i]])),
    FWD.ReverseReads = sapply(REV.orients, primerHits, fn = file.path(fn[[i]])))

  print (checkPrimers)
}

samplename <- sapply(strsplit(basename(fn), "\\."), `[`, 1)
fnF <- file.path("./", paste0(samplename, "_flt.fastq.gz"))
names(fnF) <- samplename

out <- filterAndTrim(fn, fnF, truncLen=290, maxN=0, maxEE=2, truncQ=2, rm.phix=TRUE, compress=TRUE, multithread=TRUE)

errF <- learnErrors(fnF, nbases = 1e8, multithread=TRUE, randomize=TRUE)

errPlot <- plotErrors(errF, nominalQ=TRUE)

ggsave("./errorF.png", errPlot, device="png")

dadaFs <- dada(fnF, err=errF, multithread=TRUE)

seqtab <- makeSequenceTable(dadaFs)

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)

taxa <- assignTaxonomy(seqtab.nochim, "../../taxDB/silva_nr99_v138_wSpecies_train_set.fa.gz", tryRC=TRUE,  multithread=TRUE)




